spec/unit/resolver_spec.rb

```diff
@@ -550,6 +510,152 @@ def resolve
+        end
+
+        sources = SourcesManager.sources(%w(master test_repo))
+        resolver = Resolver.new(config.sandbox, podfile, empty_graph, sources)
+        resolver.resolve
+
+        UI.warnings.should.not.match /multiple specifications/
+      end
+
+      it 'fails to resolve a dependency with an explicit source even if it can be ' \
+         'resolved using the global sources' do
+        test_repo_url = SourcesManager.source_with_name_or_url('test_repo').url
+        podfile = Podfile.new do
+          platform :ios
+          pod 'JSONKit', '1.5pre', :source => test_repo_url
+        end
+
+        sources = SourcesManager.sources(%w(master))
+        resolver = Resolver.new(config.sandbox, podfile, empty_graph, sources)
+        e = lambda { resolver.resolve }.should.raise Informative
+        e.message.should.match(/None of the spec sources contain a spec satisfying/)
+        e.message.should.match(/JSONKit/)
+        e.message.should.match(/\= 1.5pre/)
+      end
+
+      it 'resolves a dependency with an explicit source even if it can\'t be ' \
+         'resolved using the global sources' do
+        master_repo_url = SourcesManager.source_with_name_or_url('master').url
+        podfile = Podfile.new do
+          platform :ios
+          pod 'JSONKit', '1.5pre', :source => master_repo_url
+        end
+
+        sources = SourcesManager.sources(%w(test_repo))
+        sources.map(&:url).should.not.include(master_repo_url)
+        resolver = Resolver.new(config.sandbox, podfile, empty_graph, sources)
+        spec = resolver.resolve.values.flatten.first
+        spec.version.to_s.should == '1.5pre'
+        spec.defined_in_file.should == fixture('spec-repos/master/Specs/JSONKit/1.5pre/JSONKit.podspec.json')
+      end
+
+      it 'uses explicit source repos for a dependency even when it\'s transitive' do
+        master_repo_url = SourcesManager.source_with_name_or_url('master').url
+        test_repo_url = SourcesManager.source_with_name_or_url('test_repo').url
+
+        podfile = Podfile.new do
+          platform :ios
+          # KeenClient has a dependency on JSONKit 1.4
+          pod 'KeenClient', '3.2.2', :source => master_repo_url
+          pod 'JSONKit', '1.4', :source => test_repo_url
+        end
+
+        sources = SourcesManager.sources(%w(master test_repo))
+        resolver = Resolver.new(config.sandbox, podfile, empty_graph, sources)
+        resolver.resolve
+
+        possible_specs = resolver.search_for(Dependency.new('JSONKit', '1.4'))
+
+        # JSONKit, v1.4 exists in both repos, but we should only ever be offered the test_repo version.
+        possible_specs.count.should == 1
+        possible_specs.first.version.to_s.should == '1.4'
+        possible_specs.first.defined_in_file.should == fixture('spec-repos/test_repo/JSONKit/1.4/JSONKit.podspec')
+      end
+
+      it 'uses global source repos for resolving a transitive dependency even ' \
```

Comment:
Tests that for a top-level dependency with an explicit source, its transitive dependencies can be resolved by any global sources in the Podfile (unless the transitive dependency is explicitly pegged to a source or path, as tested above).

This is the same behavior that Gemfiles use, and also the one that makes the most sense to me. In most cases I think a dependency will be specified with an explicit source because it's a company internal pod. That internal pod should still be able to depend on public pods (like JSONKit or AFNetworking) without needing to clone those public pods into the internal podspec repo. If the internal pod has dependencies on other internal pods then the internal podspec source can be included at the global level as well (preferred), or those dependencies can be enumerated explicitly at the Podfile level with :source flags.